### ISSUE TITLE
JNI Layer throws Exceptions when close method is called in parallel ON DIFFERENT MODELS

### ISSUE THREAD

**Issue by ryan-deak-zefr** [VowpalWabbit/vowpal_wabbit#1300]
## Problem

In the JNI layer, when multiple passes are enabled (`> 1`) and an attempt is made to close separate models in parallel, exceptions can be thrown.  This is true even though each model has it's own lock to guard all accesses to the native code paths.  Only a **global** lock around calls to the model `close` methods seems to avoid this issue.

I'm looking for help on identifying if any critical sections of the C code that can be guarded by a lock to avoid thread-safety issues.  I'm not asking for the C code to lock.  I just want help trying to figure out where to put the locks in the Java code that wraps the C code.

### Scope

This seems to be in PR #1291 but was not fixed with PR #1295

## Discussion

From empirical testing, it appears one of these lines seems to be the problem.  I am wondering if any of these use global state.  I am trying to figure out if we can lock only over a short critical section to avoid thread-safety issues.

1. ` adjust_used_index(*vwInstance);`
1. `vwInstance->do_reset_source = true;`
1. `VW::start_parser(*vwInstance);`
1. `LEARNER::generic_driver(*vwInstance);`
1. `VW::end_parser(*vwInstance);`

## Previous Conversation
In PR #1295 there was the following conversation:

@JohnLangford 

>There should be zero shared state between multiple created VW objects.  Is
that what it's doing?  (Creating multiple distinct VW objects?)

@deaktator 

> @JohnLangford. It looks like just one VW object. Each Java call does the following on the C side:
>
> `vw* vwInstance = VW::initialize(env->GetStringUTFChars(command, NULL);`

@JohnLangford 

>A single VW object can not be operated on in multiple threads because the code inside VW is not thread safe.  If you want to have a model which is shared by multiple threads, you set this up more explicitly by initializing a new VW object with an existing model.

@deaktator

> Hey @JohnLangford. We take care of multi-threaded access to VW by locking anywhere that requires access to the C code. The thread-safety issues I encountered before were on an incomplete version of the code that locked in the wrong place. When I run the tests in parallel, they seem to work just fine now. I ran them a bunch of times with forking in the tests and didn't see any issues.

## Tracking Down What's Happening

It appears @jon-morra-zefr pretty much copied the C# code for multiple passes, so this seems like it might apply to C# as well.  Both C# and JNI C++ code appear below as well as the calling code that blows up.  

I've seen a bunch of different errors that occur at the same spot.  invalidated cache, malformed LDF feature exceptions, etc.

## Example Code That triggers exceptions

```scala
// Doing this many times in parallel with no locks causes problems.

	val vwJNI = VWLearners.create[VWTypedLearner[_]](vwLearnString)
	// Learning in here using   vwJNI.learn

	// PROBLEM AREA:
	lock.lock()     // <== NEED GLOBAL LOCKING OR EXCEPTIONS THROWN
	vwJNI.close()
	lock.unlock()   // <== NEED GLOBAL LOCKING OR EXCEPTIONS THROWN
```

### C# Code: vowpal_wabbit/cs/cli/vowpalwabbit.cpp 
```csharp
void VowpalWabbit::RunMultiPass()
{ if (m_vw->numpasses > 1)
  { try
    { adjust_used_index(*m_vw);
      m_vw->do_reset_source = true;
      VW::start_parser(*m_vw);
      LEARNER::generic_driver(*m_vw);
      VW::end_parser(*m_vw);
    }
    CATCHRETHROW
  }
}
```

### JNI C++ Code: vowpal_wabbit/java/src/main/c++/vowpalWabbit_learner_VWLearners.cc

```c
JNIEXPORT void JNICALL Java_vowpalWabbit_learner_VWLearners_performRemainingPasses(JNIEnv *env, jclass obj, jlong vwPtr)
{ try
  { vw* vwInstance = (vw*)vwPtr;
    if (vwInstance->numpasses > 1)
      { adjust_used_index(*vwInstance);
        vwInstance->do_reset_source = true;
        VW::start_parser(*vwInstance);
        LEARNER::generic_driver(*vwInstance);
        VW::end_parser(*vwInstance);
      }
  }
  catch(...)
  { rethrow_cpp_exception_as_java_exception(env);
  }
}
```

*Any thoughts?*

**Comment by JohnLangford** [comment]
A global lock seems like the obviously safe thing to do, and hence the
baseline solution.  I'm not fully following why you want to delve into
tighter locking?

Lines (1) and (2) have a race condition with line (3) since the parser,
once started, uses used_index.

-John

**Comment by jon-morra-zefr** [comment]
Our use case is that we have implemented KFold cross validation using the JNI layer for learning and predicting.  We spawn N threads that each train and evaluate a model on different slices of the same data.  If we have a global lock on the multiple passes then our learning is only parallelizable on the data ingestion which makes our whole process slower.  Is there anyway to have INDEPENDENT `vwInstance`s operate on INDEPENDENT datasets without a race condition occurring?  We're happy to lock on a small part of the code we just don't want to lock on all the passes.

**Comment by ryan-deak-zefr** [comment]
Hey @JohnLangford, the motivation for figuring this out is to allow the training of multiple models in parallel on one machine.  This can be useful for things like parallelizing cross validation.  This issue seems to have popped up for @jon-morra-zefr when training N individual JNI models on N threads, where one model is only ever accessed by *its own* thread.  The training of the first pass has no issues. Training the subsequent passes is the issue.

**Comment by ryan-deak-zefr** [comment]
@jon-morra-zefr it seems we have a race condition in updating this issue.

**Comment by JohnLangford** [comment]
Is the claim that a per-object lock does not work?  Or is not efficient
enough?  (When I said "global", I meant "per-object".)

-John

**Comment by deaktator** [comment]
The claim is that per-object locks are not sufficient to prevent exceptions being thrown.  **One global lock** across all models seems to always work.

**Comment by ryan-deak-zefr** [comment]
I think this has something do do with caching.  The problem seems to come after a message like:

```
WARNING: cache file is ignored as it's made with less bit precision than required!
creating cache_file = /tmp/123.vw.cache
```

I tried this on scalar models and CSOAA models.  Both seem to have the issue.  

Attached is a unit test that shows the issues.
[MultiModelConcurrentCloseTest.java.zip](https://github.com/JohnLangford/vowpal_wabbit/files/1178303/MultiModelConcurrentCloseTest.java.zip)

**Comment by JohnLangford** [comment]
There should never be a problem with a per-object lock as far as the code
itself goes.

You can certainly have a race condition w.r.t. cache files.   Often we get
around this by creating a cache file once then using it as a read-only
object.

If the cache file is created with the largest number of bits that you want,
VW will downsample bits as necessary for lesser values of -b.

-John

**Comment by JohnLangford** [comment]
Is cache file creation the problem?

**Comment by deaktator** [comment]
@JohnLangford: I think it's a caching issue, not 100% positive. I wasn't setting -b so I am assuming it's 18.  I don't create the cache file, just pass a path to VW. Do I need to actually create it or does VW take care of that when specifying --cache_file?

**Comment by JohnLangford** [comment]
VW creates the cache file if it does not exist and otherwise uses it.

**Comment by deaktator** [comment]
So if I never create or access the file on the Java side and don't *explicitly* call any C code that does it, do you have any idea what could lead to the cache file getting screwed up or corrupted? I looked at the parser code but couldn't figure it out.

**Comment by JohnLangford** [comment]
If you don't use -c or --cache_file in the C code, then no cache file is
created or used by VW.  The system will however complain if you have no
cache file and try to use --passes.

**Comment by JohnLangford** [comment]
So, does this problem persist when using distinct cachefiles?  

Similar to the cachefile issue, could there be a race condition in model writing?

**Comment by jon-morra-zefr** [comment]
I just checked and it looks like writing different models concurrently is working from the JNI.  Multiple passes is not working.

**Comment by JohnLangford** [comment]
How can the bug be isolated further?  You have verified that different
cachefiles are being used in the different objects, right?

**Comment by jon-morra-zefr** [comment]
Yes, we have verified that.  @ryan-deak-zefr submitted code that reproduces this bug.  After the incorporation of the PR #1303 this code will no longer create the bug, but that's just cause I have masked the problem with a global lock.

**Comment by JohnLangford** [comment]
The existing C++ codebase was audited by @eisber to remove all static variables (i.e. those that would be shared by default amongst multiple instances).

**Comment by cpdomina** [comment]
@ryan-deak-zefr does the errors happen if you use the `-k` parameter? From what I understand in the documentation, it makes sure that cache files are not reused.

**Comment by jon-morra-zefr** [comment]
It still does happen.

**Comment by Wirack** [comment]
I have experienced similar problems with the c# binding when running multipass in parallel on different objects with different cachefiles.

**Comment by JohnLangford** [comment]
This is a frustrating bug to track down.  Ideas on how to find it?  Apparently, there is some shared global state, but I have no idea where it is at present.

**Comment by Wirack** [comment]
It might be something related to caching. When I run [RunMultiPass()](https://github.com/JohnLangford/vowpal_wabbit/blob/master/cs/cli/vowpalwabbit.h#L59) in parallel (for k-fold crossvalidation) I sporadically receive [argh, a bug in caching of some sort](https://github.com/JohnLangford/vowpal_wabbit/blob/ec54aa3bf9c263e0d314140ba7c44ab6ae45f181/vowpalwabbit/parser.cc#L272).

 I use oaa with 27 bits and also use quadratic features and n-skip-k-grams.

**Comment by JohnLangford** [comment]
Is this a shared cache that is not precomputed?  if so... you shouldn't do
that.

-John

**Comment by Wirack** [comment]
Each VW instance hasassigned a separate cachefilename.

**Comment by JohnLangford** [comment]
Hmm---it should work then.

Do you have some minimal piece of code which reliably excites the bug?

-John

**Comment by Wirack** [comment]
On one machine I was able to reproduce the bug with the following code: 

```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using VW;
namespace VowpalWabbiterror
{
	class Program
	{
		static void Main(string[] args)
		{
			//Setup some vw-instances
			List<VowpalWabbit> vwInstances = new List<VowpalWabbit>();
			for(int i = 0; i < 100; i++)
			{
				vwInstances.Add(new VowpalWabbit($"--oaa 3 --cache_file {i}.cache -k --early_terminate 10 --passes 10"));
			}

			string ex1 = "1 | some features here",
				   ex2 = "2 | some other features",
				   ex3 = "3 | some more";

			//Prepare some data
			for(int i = 0; i < 10000; i++)
			{
				foreach (var vw in vwInstances) {
					vw.Learn(ex1);
					vw.Learn(ex2);
					vw.Learn(ex3);
				}
			}

			Parallel.ForEach(vwInstances, vw => { vw.RunMultiPass(); vw.Dispose(); });
		

		}
	}
}

```

**Comment by Wirack** [comment]
Before receiving  `argh, a bug in caching of some sort` I get the following console output:

```
truncated example! wanted: 272730423296 bytes
truncated example! 0 10
truncated example! 0 10
truncated example! wanted: 272730423296 bytes
truncated example! 0 10
truncated example! 0 10
truncated example! wanted: 272730423296 bytes
truncated example! wanted: 272730423296 bytes
truncated example! 0 10
truncated example! 0 10
truncated example! wanted: 272730423296 bytes
truncated example! 0 10
truncated example! wanted: 272730423296 bytes
truncated example! wanted: 17873661021126656 bytes
truncated example! 0 10
truncated example! 0 10
truncated example! 0 10
truncated example! 0 10
truncated example! 0 10
truncated example! 0 10
truncated example! 0 10
truncated example! 0 10
truncated example! 0 10
truncated example! 0 10
label 593 is not in {1,3} This won't work right.
truncated example! 0 10
truncated example! wanted: 272730423296 bytes
truncated example! 0 10
truncated example! 0 10
truncated example! 0 10
truncated example! 0 10
truncated example! 0 10
truncated example! 0 10
truncated example! wanted: 17873661021126656 bytes
truncated example! wanted: 16140901064495857664 bytes
truncated example! wanted: 17873661021126656 bytes
truncated example! wanted: 272730423296 bytes
truncated example! 0 10
truncated example! 0 10
truncated example! wanted: 17873661021126656 bytes
cache has possibly incompatible version, rebuilding
cache has possibly incompatible version, rebuilding
truncated example! wanted: 272730423296 bytes
```

**Comment by jon-morra-zefr** [comment]
I think that this should be closed as it is "fixed" in the JNI code.  I use quotes because the fix is to use a global lock.  Should the native C code be fixed a new issue should be opened to remove this global lock from the JNI layer to make it more performant.

**Comment by ryan-deak-zefr** [comment]
@JohnLangford @jon-morra-zefr @Wirack, I agree with @jon-morra-zefr.  I think #1300 and #1303 should be closed and we should open an issue linking to these issues.  I think people will care more to find this stuff if we show it's an issue in the base C library.  If we translate [@Wirack's example](https://github.com/JohnLangford/vowpal_wabbit/issues/1300#issuecomment-347231627) to C, do we get similar behaviors to the ones found in the JNI and C# interfaces?  Is expected to #1338 fix this?  I can test this once the changes in #1338 are merged.

**Comment by JohnLangford** [comment]
This one has been persistent.  I nailed a separate heisenbug recently (unlikely to be related to this).  What I'm missing is an easy repro to tackle this myself.

**Comment by jmorra** [comment]
The easiest way to verify if this works or not is to remove the global lock and run the code from the Java side with multiple passes.  If you have reason to believe this will be fixed by now I'm happy to try it out.

**Comment by JohnLangford** [comment]
I have no reason to believe it is fixed, but on the other hand I don't have an easy/sure way to reproduce it, hence no progress.  If you can make the equivalent of #1383 for this issue it will substantially improve the odds of debugging.

**Comment by pferrel** [comment]
@jon-morra-zefr in https://github.com/JohnLangford/vowpal_wabbit/issues/1300#issuecomment-347252617 says the JNI part is fixed. What version, should we get master or a release? The global lock is in JNI but does work around the C++ problem? If so we should not see the JVM become wedged, assuming it is the problem we are seeing?

**Comment by pferrel** [comment]
One further note, using 8.4.1 we do the following when deleting an instance of VW JNI and this discussion makes me worry that the VW code may not return synchronously.

```
vw.close
    try {
      Await.result(
        actors.terminate().map { _ =>
          logger.info(s"Closing vw $vw")
          if (vw != null.asInstanceOf[VWMulticlassLearner]) vw.close()
        }.map { _ =>
          if (Files.exists(Paths.get(modelPath)) && !Files.isDirectory(Paths.get(modelPath))) {
            logger.info(s"Deleting file ${modelPath} for engine $engineId and vw $vw")
            while (!Files.deleteIfExists(Paths.get(modelPath))) {
              logger.info(s"trying to delete ${modelPath}")
            }
          } else logger.info(s"$vw of engine $engineId has no file to delete")
        }, 2 minutes)
      logger.info(s"Algorithm for $engineId (created with params $params) was destroyed")
    } catch {
      case e: IOException => logger.error(s"Can't destroy $this", e)
      case e: TimeoutException => logger.error(s"Error unable to delete the VW model file for $engineId at $modelPath in the 3 second timeout.", e)
    }
    try {
      Await.result(
        actors.terminate().map { _ =>
          logger.info(s"Closing vw $vw")
          if (vw != null.asInstanceOf[VWMulticlassLearner]) vw.close()
        }.map { _ =>
          if (Files.exists(Paths.get(modelPath)) && !Files.isDirectory(Paths.get(modelPath))) {
            logger.info(s"Deleting file ${modelPath} for engine $engineId and vw $vw")
            while (!Files.deleteIfExists(Paths.get(modelPath))) {
              logger.info(s"trying to delete ${modelPath}")
            }
          } else logger.info(s"$vw of engine $engineId has no file to delete")
        }, 2 minutes)
      logger.info(s"Algorithm for $engineId (created with params $params) was destroyed")
    } catch {
      case e: IOException => logger.error(s"Can't destroy $this", e)
      case e: TimeoutException => logger.error(s"Error unable to delete the VW model file for $engineId at $modelPath in the 3 second timeout.", e)
    }
```

This tries to close then delete the model before the object is garbage collected and removed. 

Is the vw.close method synchronous? Meaning it is done with the model file after return. If not we may be deleting a file being modified?

Something is happening at close & delete that wedges the JVM, we are having a hell of a time tracing it.

**Comment by deaktator** [comment]
@pferrel. Did you try the 8.5.0 tag? @JohnLangford @jmorra, do we want to release 8.5.0 vw jni?

**Comment by pferrel** [comment]
@deaktator my question is which JNI to use to get the lock work-around to the subject of this issue, which would seem to be similar to a global lock around vw.close.

There is no 8.4.1 tag or branch even though there is a maven-verse release called vw-jni-8.4.1. so not really sure what is in that release.  The 8.4.0 tag JNI code has lock usage in VWBase: https://github.com/JohnLangford/vowpal_wabbit/blob/8.4.0/java/src/main/java/vowpalWabbit/learner/VWBase.java#L27

**Comment by pferrel** [comment]
@deaktator @jmorra @JohnLangford My other question is after the return of vw.close() is it safe to immediately delete the model file? We have some very hard to debug hang of the JVM related to delete, which closes first then removes the model file.

**Comment by ryan-deak-zefr** [comment]
@pferrel: If you look at the bottom of the [JNI README.md](https://github.com/JohnLangford/vowpal_wabbit/blob/master/java/README.md), you'll see the git hash for the 8.4.1 JNI release (hash = `10bd09ab06f59291e04ad7805e88fd3e693b7159`).  If you checkout that hash and look at the code for `VWBase`, you should see it has a global lock.

```bash
git clone https://github.com/JohnLangford/vowpal_wabbit.git
cd vowpal_wabbit
git checkout 10bd09ab06f59291e04ad7805e88fd3e693b7159
cat java/src/main/java/vowpalWabbit/learner/VWBase.java
```

HAS:

```java
    final public void close() throws IOException {
        try {
            closer().call();
        }
        catch (Exception e) {
            throw new IOException("An exception occurred while attempting to close VW Model with native pointer " +
                                  nativePointer, e);
        }
    }

    @Override
    final public Callable<Boolean> closer() {
        return new Closer();
    }

    private class Closer implements Callable<Boolean> {
        private Closer() {}
        @Override
        public Boolean call() {
            lock.lock();
            try {
                final boolean attemptingToClose = isOpen;
                if (isOpen) {
                    isOpen = false;
                    VWBase.globalLock.lock();
                    try {
                        VWLearners.performRemainingPasses(nativePointer);
                    }
                    finally {
                        globalLock.unlock();
                    }
                    VWLearners.closeInstance(nativePointer);
                }
                return attemptingToClose;
            }
            finally {
                lock.unlock();
            }
        }
    }
```

Therefore, the `close()` function is blocking and when it returns, it *should (presumably)* be safe to delete the model file.  It also has the global lock so it should be OK at not barfing when concurrently calling close on a bunch of models.

**Comment by pferrel** [comment]
@deaktator Awesome! That is what I was coming to believe. We do have some problem with deleting instances but will look elsewhere. Thanks.

**Comment by pferrel** [comment]
@deaktator BTW is it safe to do nothing on killing the JVM with `kill -2`? Should we hook the shutdown and attempt to vw.close() all instances? I know wrong place for the question but hey, it's remotely related :-)

**Comment by JohnLangford** [comment]
This one still hasn't been fixed after a significant amount of time.  How do we get it fixed?  I'm willing to set aside some time to work with the java experts to work this out.

**Comment by matanster** [comment]
This one is not fixed in the freshly announced 8.6.0, [or is it](https://github.com/JohnLangford/vowpal_wabbit/issues/1486)?

**Comment by JohnLangford** [comment]
Not fixed explicitly---I don't have a repro.

**Comment by deaktator** [comment]
@JohnLangford we should do a release of jni 8.6.0. I'll talk to Jon.

**Comment by jackgerrits** [comment]
@kumpera

**Comment by kumpera** [comment]
I debugger the example from https://github.com/VowpalWabbit/vowpal_wabbit/issues/1300#issuecomment-347231627 and I can reproduce it with master.

The issue is that each instance of VowpalWabbit is trying to use stdin and introducing a race condition as they close it.

While we're working on a fix, there's a workaround, which is to pass `--stdin_off` as your first VW argument.
